Update "PWAs on Oculus 2" blog post (#9051)
GoogleChrome/web.dev: src/site/content/en/blog/pwas-on-oculus-2/index.md

</table>
</div>

### Packaging PWAs with Bubblewrap CLI

[Bubblewrap](https://github.com/GoogleChromeLabs/bubblewrap) is an open source set of libraries and
a command line tool (CLI) for Node.js developed by the Google Chrome team that helps developers

> generate, build and sign an Android project that launches your PWA as a

Reviewer: ```suggestion
generate, build, and sign an Android project that launches your PWA as a
```

> [Bubblewrap](https://github.com/GoogleChromeLabs/bubblewrap) is an open source set of libraries and
> a command line tool (CLI) for Node.js developed by the Google Chrome team that helps developers
> generate, build, and sign an Android project that launches your PWA as a
> [Trusted Web Activity](https://developer.chrome.com/docs/android/trusted-web-activity/) (TWA).

Reviewer 2: ```suggestion
[Bubblewrap](https://github.com/GoogleChromeLabs/bubblewrap) is an open source set of libraries and
a command line tool (CLI) for Node.js. Bubblewrap is developed by the Google Chrome team, to help developers
generate, build, and sign an Android project that launches your PWA as a
[Trusted Web Activity](https://developer.chrome.com/docs/android/trusted-web-activity/) (TWA).
```

Meta Quest Browser currently doesn't fully support TWA, but starting from version 1.18.0,

> Bubblewrap [supports](https://twitter.com/alexey_rodionov/status/1544374516767285249) packaging

Reviewer: ```suggestion
Bubblewrap [supports](https://github.com/GoogleChromeLabs/bubblewrap/releases/tag/v1.18.0) packaging
```

PWAs for Meta Quest devices.

> It can generate *universal* APKs that open TWA on regular Android devices and open Meta Quest

Reviewer: ```suggestion
It can generate *universal* APKs that open a TWA on regular Android devices and open Meta Quest
```

> Browser on Meta Quest devices, and can be published to Google Play Store and Meta Quest Store.

Reviewer: ```suggestion
Browser on Meta Quest devices, so it can be published to the Google Play Store and the Meta Quest Store.
```

> It can generate *universal* APKs that open a TWA on regular Android devices and open Meta Quest
> Browser on Meta Quest devices, so it can be published to the Google Play Store and the Meta Quest Store.

Reviewer 2: ```suggestion
It can generate *universal* APKs that open a TWA on regular Android devices and Meta Quest
Browser on Meta Quest devices, so the APK can be published to the Google Play Store and the Meta Quest Store.
```

Reviewer 2: I think this is right, it wasn't clear what "it" referred to.

Author: @rachelandrew Thanks! Applied. :)

> Assuming you have [Node.js](https://nodejs.org) installed. Bubblewrap CLI can be installed with the

Reviewer: ```suggestion
Assuming you have [Node.js](https://nodejs.org) installed, Bubblewrap CLI can be installed with the
```

following command:

```bash
npm i -g @bubblewrap/cli
```

When running Bubblewrap for the first time, it will offer to automatically download and install the

> required external dependencies — Java Development Kit (JDK) and Android SDK Build Tools.

Reviewer: ```suggestion
required external dependencies—Java Development Kit (JDK) and Android SDK Build Tools.
```

> To generate a Meta Quest compatible Android project that wraps your PWA just run the `init` command

Reviewer: ```suggestion
To generate a Meta Quest compatible Android project that wraps your PWA, run the `init` command
```

with the `--metaquest` flag and follow the wizard:

```bash
bubblewrap init --manifest="https://your.web.app/manifest.json" --metaquest
```

Once the project has been generated, build and sign it by running:

```bash
bubblewrap build
```

This will output a file called `app-release-signed.apk`. This file can be installed on the device

> or uploaded to Meta Quest Store and Google Play Store.

Reviewer: ```suggestion
or uploaded to the Meta Quest Store and the Google Play Store.
```

> {% Aside %} Pro tip: you can also use the `--chromeosonly` flag in addition to the `--metaquest`

Reviewer: ```suggestion
{% Aside 'important' %} You can also use the `--chromeosonly` flag in addition to the `--metaquest`
```

flag to make the APK compatible not only with Meta Quest and regular Android devices, but also with
[ChromeOS](https://chromeos.dev) devices. {% endAside %}

### Packaging PWAs with Oculus Platform Utility

[Oculus Platform Utility](https://developer.oculus.com/documentation/web/pwa-packaging/#download-the-cli)
is the official command line tool developed by Meta for publishing apps for Oculus Rift and Meta
Quest devices.

It also allows to package PWAs for Meta Quest devices with the `create-pwa` command and publish
them to Meta Quest Store and App Lab.

Set the output file name via the `-o` parameter and the path to Android SDK via the `--android-sdk`
parameter.

Point the tool at the live URL of the web app manifest via the `--web-manifest-url` parameter.

If you don't have a manifest on your live PWA or wish to override the live manifest, you can still
generate an APK for your PWA using a local manifest file and the `--manifest-content-file`
parameter.

> To leave the manifest as pure as possible, use the `--package-name` parameter with a value in
> reverse domain name notation (e.g., `com.company.app.pwa`), rather than adding the proprietary
> `ovr_package_name` field to the manifest.

Reviewer 2: ```suggestion
To leave the manifest as pure as possible, use the `--package-name` parameter with a value in
reverse domain name notation (for example, `com.company.app.pwa`), rather than adding the proprietary
`ovr_package_name` field to the manifest.
```

```bash
ovr-platform-util create-pwa -o output.apk --android-sdk ~/bin/android-10 --manifest-content-file manifest.json --package-name com.company.app.pwa
```

{% Aside 'caution' %}
APK files generated by `ovr-platform-util` are only compatible with Meta Quest devices and cannot
be run on regular Android devices.
{% endAside %}

### Packaging PWAs with PWABuilder

> Using [PWABuilder](https://pwabuilder.com) is the easiest and recommended way to package PWAs for

Reviewer: ```suggestion
Using [PWABuilder](https://pwabuilder.com) is in my view the easiest and therefore recommended way to package PWAs for
```

Author: @tomayac This text hasn't been changed as part of this pull request. This is not my point of view, this is the official statement of Meta: https://developer.oculus.com/documentation/web/pwa-building-with-pwabuilder/. :wink:

Reviewer: Still better not to say something about others' projects that could be interpreted the wrong way.

Meta Quest at the moment.

> PWABuilder is an [open source](https://github.com/pwa-builder), community guided project founded by

Reviewer: ```suggestion
PWABuilder is an [open source](https://github.com/pwa-builder) project founded by
```

Author: @tomayac This text hasn't been changed as part of this pull request. The "community guided" part has been taken from the https://pwabuilder.com footer: "PWABuilder was founded by Microsoft as a community guided, open source project to help move PWA adoption forward."

Reviewer: Still better not to say something about others' projects that could be interpreted the wrong way.

Author: @tomayac You can also find this phrase at https://developer.oculus.com/documentation/web/pwa-building-with-pwabuilder/. :wink:

Reviewer: Still better not to say something about others' projects that could be interpreted the wrong way.

> Using [PWABuilder](https://pwabuilder.com) is in my view the easiest and therefore recommended way to package PWAs for

Reviewer: ```suggestion
Using [PWABuilder](https://pwabuilder.com) is in the authors' view the easiest and therefore recommended way to package PWAs for
```

Microsoft, that allows developers to package and sign their PWAs for publishing to various stores,
including Microsoft Store, Google Play Store, App Store, and Meta Quest Store.

Packaging PWAs with PWABuilder is as easy as entering the URL of a PWA, selecting the platform
(**Meta Quest** in this case), entering/editing metadata for the app, and clicking the **Generate**
button.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/sWjB34HjwjatCMUz39Q3.png", alt="Packaging PWAs for Meta Quest with PWABuilder.", width="800", height="450" %}

{% Aside %} Pro tip: since PWABuilder is a PWA itself that works everywhere there is a browser, you
can open it in the Meta Quest Browser and perform all the steps right on your Meta Quest.
{% endAside %}

PWABuilder uses the `ovr-platform-util` under the hood to package PWAs for Meta Quest, but the size

> is [smaller up to 1000x](https://twitter.com/alexey_rodionov/status/1547301992925540352).

Reviewer: ```suggestion
can be [smaller](https://github.com/pwa-builder/pwabuilder-oculus/pull/7) due to dead code removal.
```

### Installing PWAs with ADB

> After creating the APK, developers can sideload it to the Meta Quest device using Android Debug

Reviewer: ```suggestion
After creating the APK, developers can sideload it to the Meta Quest device using the Android Debug
```

Bridge (ADB) via USB or Wi-Fi:

```bash
adb install app-release-signed.apk
```

If you use Bubblewrap CLI for packaging PWAs, it provides a convenient alias command to sideload
the APK:

```bash
bubblewrap install
```

{% Aside %} Pro tip: you can use [Android Web Toolbox](https://yume-chan.github.io/ya-webadb/install)
to sideload the APK. It's an [open source](https://github.com/yume-chan/ya-webadb/) web app that allows
you to access all ADB functionality right from the browser even from another Android device. No
installation or drivers are required thanks to the [Web USB API](/usb/). {% endAside %}

To enable ADB on the Meta Quest device, you must
[enable developer mode](https://developer.oculus.com/documentation/native/android/mobile-device-setup/).

Sideloaded apps appear in an **Unknown Sources** section in the app drawer.

### App submission

{% Aside %} Submission and consideration for the Oculus Store is only available if you have been
approved as an Oculus Quest Store developer. Distribution of PWAs via
[App Lab](https://developer.oculus.com/blog/introducing-app-lab-a-new-way-to-distribute-oculus-quest-apps/)

> is not currently available. The Oculus team will share more on when and how you can submit a PWA to

Reviewer: ```suggestion
is not currently available. The Meta team will share more on when and how you can submit a PWA to
```

(Not sure about this one. Maybe it's fine to leave Oculus.)
